modules/imgproc/src/filter.cpp

```diff
@@ -4555,222 +4528,577 @@ cv::Ptr<cv::FilterEngine> cv::createLinearFilter( int _srcType, int _dstType,
+        Mat dst(Size(width, height), dst_type, dst_data, dst_step);
         Mat temp;
+        int src_channels = CV_MAT_CN(src_type);
+        int dst_channels = CV_MAT_CN(dst_type);
+        int ddepth = CV_MAT_DEPTH(dst_type);
         // crossCorr doesn't accept non-zero delta with multiple channels
-        if( src.channels() != 1 && delta != 0 )
-        {
+        if (src_channels != 1 && delta != 0) {
             // The semantics of filter2D require that the delta be applied
             // as floating-point math.  So wee need an intermediate Mat
             // with a float datatype.  If the dest is already floats,
             // we just use that.
-            int corrDepth = dst.depth();
-            if( (dst.depth() == CV_32F || dst.depth() == CV_64F) &&
-                src.data != dst.data )
-            {
-                temp = dst;
+            int corrDepth = ddepth;
+            if ((ddepth == CV_32F || ddepth == CV_64F) && src_data != dst_data) {
+                temp = Mat(Size(width, height), dst_type, dst_data, dst_step);
+            } else {
+                corrDepth = ddepth == CV_64F ? CV_64F : CV_32F;
+                temp.create(Size(width, height), CV_MAKETYPE(corrDepth, dst_channels));
             }
-            else
-            {
-                corrDepth = dst.depth() == CV_64F ? CV_64F : CV_32F;
-                temp.create( dst.size(), CV_MAKETYPE(corrDepth, dst.channels()) );
+            crossCorr(src, kernel, temp, src.size(),
+                      CV_MAKETYPE(corrDepth, src_channels),
+                      anchor, 0, borderType);
+            add(temp, delta, temp);
+            if (temp.data != dst_data) {
+                temp.convertTo(dst, dst.type());
             }
-            crossCorr( src, kernel, temp, src.size(),
-                       CV_MAKETYPE(corrDepth, src.channels()),
-                       anchor, 0, borderType );
-            add( temp, delta, temp );
-            if ( temp.data != dst.data )
-            {
-                temp.convertTo( dst, dst.type() );
-            }
-        }
-        else
-        {
-            if( src.data != dst.data )
-                temp = dst;
+        } else {
+            if (src_data != dst_data)
+                temp = Mat(Size(width, height), dst_type, dst_data, dst_step);
             else
-                temp.create(dst.size(), dst.type());
-            crossCorr( src, kernel, temp, src.size(),
-                       CV_MAKETYPE(ddepth, src.channels()),
-                       anchor, delta, borderType );
-            if( temp.data != dst.data )
+                temp.create(Size(width, height), dst_type);
+            crossCorr(src, kernel, temp, src.size(),
+                      CV_MAKETYPE(ddepth, src_channels),
+                      anchor, delta, borderType);
+            if (temp.data != dst_data)
                 temp.copyTo(dst);
         }
-        return;
+    }
+};
+
+struct OcvFilter : public HalFilterImpl
```

Comment:
as I see, you are using "Impl" paradigm anyway, so it's another reason to use OOP style for the HAL